src/AWSCredentials.jl

```diff
@@ -56,392 +58,460 @@ mutable struct AWSCredentials
+    end
 
-    String(http_get("http://169.254.169.254/latest/meta-data/$key").body)
+    return nothing
 end
 
-function instance_credentials()
+
+function _get_awscreds_instance()
     if haskey(ENV, "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
         return ecs_instance_credentials()
     elseif localhost_maybe_ec2()
         return ec2_instance_credentials()
-    else
-        return nothing
     end
+
+    return nothing
 end
 
+
 """
-Load [Instance Profile Credentials]
-(http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html#instance-metadata-security-credentials)
-for EC2 virtual machine.
+    ec2_instance_credentials() -> AWSCredentials
+
+Parse the EC2 metadata to retrieve AWSCredentials.
 """
 function ec2_instance_credentials()
+    info  = _ec2_metadata("iam/info")
+    info  = JSON.parse(info)
 
-    @assert localhost_maybe_ec2()
-
-    info  = ec2_metadata("iam/info")
-    info  = LazyJSON.value(info)
-
-    name  = ec2_metadata("iam/security-credentials/")
-    creds = ec2_metadata("iam/security-credentials/$name")
-    new_creds = LazyJSON.value(creds)
-
-    if debug_level > 0
-        print("Loading AWSCredentials from EC2 metadata... ")
-    end
+    name  = _ec2_metadata("iam/security-credentials/")
+    creds = _ec2_metadata("iam/security-credentials/$name")
+    new_creds = JSON.parse(creds)
 
     expiry = DateTime(strip(new_creds["Expiration"], 'Z'))
 
-    AWSCredentials(new_creds["AccessKeyId"],
-                   new_creds["SecretAccessKey"],
-                   new_creds["Token"],
-                   info["InstanceProfileArn"];
-                   expiry = expiry)
+    return AWSCredentials(
+        new_creds["AccessKeyId"],
+        new_creds["SecretAccessKey"],
+        new_creds["Token"],
+        info["InstanceProfileArn"];
+        expiry=expiry,
+        renew=ec2_instance_credentials
+    )
 end
 
 
 """
-Load [ECS Task Credentials]
-(http://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html)
+    ecs_instance_credentials() -> Union{AWSCredential, Nothing}
+
+Retrieve credentials from the local endpoint. Return `nothing` if not running on an ECS
+instance.
+
+More information can be found at:
+https://docs.aws.amazon.com/AmazonECS/latest/developerguide/task-iam-roles.html
+
+# Returns
+- `AWSCredentials`: AWSCredentials from `ECS` credentials URI, `nothing` if the Env Var is
+    not set (not running on an ECS container instance)
+
+# Throws
+- `StatusError`: If the response status is >= 300
+- `ParsingError`: Invalid HTTP request target
 """
 function ecs_instance_credentials()
-
-    @assert haskey(ENV, "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+    if !haskey(ENV, "AWS_CONTAINER_CREDENTIALS_RELATIVE_URI")
+        return nothing
+    end
 
     uri = ENV["AWS_CONTAINER_CREDENTIALS_RELATIVE_URI"]
 
-    new_creds = String(http_get("http://169.254.170.2$uri").body)
+    response = @mock HTTP.request("GET", "http://169.254.170.2$uri")
```

Comment:
`HTTP.request` is not a valid replacement for `http_get`. `http_get` does much more, including retries and adding specific headers.

Reply:
Same thing on line 260